pkg/storage/stores/shipper/downloads/table.go

```diff
@@ -51,449 +49,270 @@ type Table struct {
-			continue
-		}
-		boltdb, err := shipper_util.SafeOpenBoltdbFile(filePath)
-		if err != nil {
-			return err
-		}
-
-		var stat os.FileInfo
-		stat, err = os.Stat(filePath)
-		if err != nil {
-			return err
-		}
-
-		totalFilesSize += stat.Size()
-
-		t.dbs[file.Name] = boltdb
+		return nil, err
 	}
 
-	duration := time.Since(startTime).Seconds()
-	t.metrics.tablesDownloadDurationSeconds.add(t.name, duration)
-	t.metrics.tablesDownloadSizeBytes.add(t.name, totalFilesSize)
-	level.Debug(spanLogger).Log("total-files-size", totalFilesSize)
+	table.indexSets[""] = commonIndexSet
 
-	return
+	return &table, nil
 }
 
-// Closes references to all the dbs.
+// Close Closes references to all the dbs.
 func (t *Table) Close() {
-	// stop the initialization if it is still ongoing.
-	t.cancelFunc()
-
-	t.dbsMtx.Lock()
-	defer t.dbsMtx.Unlock()
+	t.indexSetsMtx.Lock()
+	defer t.indexSetsMtx.Unlock()
 
-	for name, db := range t.dbs {
-		if err := db.Close(); err != nil {
-			level.Error(util_log.Logger).Log("msg", fmt.Sprintf("failed to close file %s for table %s", name, t.name), "err", err)
-		}
+	for _, userIndexSet := range t.indexSets {
+		userIndexSet.Close()
 	}
 
-	t.dbs = map[string]*bbolt.DB{}
+	t.indexSets = map[string]IndexSet{}
 }
 
 // MultiQueries runs multiple queries without having to take lock multiple times for each query.
 func (t *Table) MultiQueries(ctx context.Context, queries []chunk.IndexQuery, callback chunk_util.Callback) error {
-	// let us check if table is ready for use while also honoring the context timeout
-	select {
-	case <-ctx.Done():
-		return ctx.Err()
-	case <-t.ready:
-	}
-
-	t.dbsMtx.RLock()
-	defer t.dbsMtx.RUnlock()
-
-	if t.err != nil {
-		return t.err
+	userID, err := tenant.TenantID(ctx)
+	if err != nil {
+		return err
 	}
 
-	t.lastUsedAt = time.Now()
-
-	logger := util_log.WithContext(ctx, util_log.Logger)
-
-	level.Debug(logger).Log("table-name", t.name, "query-count", len(queries))
-
-	for name, db := range t.dbs {
-		err := db.View(func(tx *bbolt.Tx) error {
-			bucket := tx.Bucket(bucketName)
-			if bucket == nil {
-				return nil
-			}
-
-			for _, query := range queries {
-				if err := t.boltDBIndexClient.QueryWithCursor(ctx, bucket.Cursor(), query, callback); err != nil {
-					return err
-				}
-			}
-
-			return nil
-		})
+	for _, uid := range []string{userID, ""} {
+		indexSet, err := t.getOrCreateIndexSet(uid, true)
 		if err != nil {
 			return err
 		}
 
-		level.Debug(logger).Log("queried-db", name)
-	}
+		if indexSet.Err() != nil {
+			level.Error(util_log.WithContext(ctx, t.logger)).Log("msg", fmt.Sprintf("index set %s has some problem, cleaning it up", uid), "err", indexSet.Err())
+			if err := indexSet.DropAllDBs(); err != nil {
+				level.Error(t.logger).Log("msg", fmt.Sprintf("failed to cleanup broken index set %s", uid), "err", err)
+			}
 
-	return nil
-}
+			t.indexSetsMtx.Lock()
```

Comment:
Since this lock is taken with the sync should we move that to another go routine, or make that index to be remove later ?

Reply:
Sorry, didn't get you here. Do you mean removing broken indexSet somewhere else and not at query time?

Reply:
Yeah, I think it's fine to set them aside or ask a worker pool to do this. I don't like the lock basically.

Reply:
Sure you don't have to do it now let's see how it goes